---
title: "MEFISTO Analysis"
author: "Cyrus A. Mallon"
date: "16 July 2024"
output:
  html_document:
    toc: true
    df_print: paged
  chunk_output_type: console
  pdf_document:
    toc: true
    number_sections: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
# Set global chunk options
knitr::opts_chunk$set(echo = TRUE, results = 'console')
```

# Intro
This is the downstream plotting and analysis, showcasing the results of the MEFISTO analysis. Factor trajectories are first plotted, then their scatter plots, and finally the weights that most influence each trajectory.

## Load and/or Install Necessary Packages
```{r, echo=FALSE}

# install and load packages ----

# list of packages
packages <- c("here", "tidyverse", "MOFA2", "patchwork", "magrittr")

# if package is missing, then install
install_if_missing <- function(pkg) {
  if (!requireNamespace(pkg, quietly = TRUE)) {
    if (pkg == "MOFA2") {
      if (!requireNamespace("BiocManager", quietly = TRUE)) {
        install.packages("BiocManager")
      }
      BiocManager::install("MOFA2")
    } else {
      install.packages(pkg, dependencies = TRUE)
    }
  }
}

# install packages ----
for (pkg in packages) {
  install_if_missing(pkg)
}

# load packages----
for (pkg in packages) {
  library(pkg, character.only = TRUE)
}
```

## Load Needed Data
Make sure to change directories if running on your own machine.
```{r, warning=FALSE, message=FALSE}
# load model and data ----
load(here("MOFAobject_untrained_26_01_2024.rda"))
load(here("feature_matrix_26_01_2024.rda"))
load(here("data_for_MOFA_26_01_2024.rda"))

MOFAobject <- load_model(here("LLNEXT_microbiome_model_26_01_2024.hdf5"), load_interpol_Z = FALSE)
```

## Plot Variance Explained By MEFISTO Model
Here I am plotting total variance explained by all factors together and separately.
```{r}
# plot variance explained ----
res_variance <- calculate_variance_explained(MOFAobject)

# how many factors do I want to plot?
start_factor <- 1
end_factor <- 10

# create a vector of factor names
factor_names <- paste0("factor_", seq(start_factor, end_factor))

# plot as bar charts
p_variance_bar_charts <-
as_tibble(res_variance$r2_per_factor$single_group) %>% 
  mutate(factor = all_of(factor_names)) %>%
  mutate(factor = factor(factor, levels = factor_names )) %>%
  ggplot() +
  geom_bar(aes(x = factor, y = microbiome), stat = "identity") +
  ylab("variance explained (%)") +
  xlab("factor") +
  ggtitle("Variance Explained by Factor") +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5)) 
  

# examine variance total
p <- plot_variance_explained(MOFAobject, plot_total = T)

# variance decomposition by factor
p_variance_by_factor <-
p[[1]] + theme(axis.text.x = element_text(angle = 90))

# total variance explained per view 
p_variance_total <-
p[[2]] + theme(axis.text.x = element_text(angle = 90))

# plot projections
p_dimensions_reduced <-
plot_factors(MOFAobject, color_by = "time")
```

## Prep data to look at factor trajectories
Here factor trajectories are plotting along the most significant hit from the post-hoc association analysis of high priority phenotypes. (N.B. I excluded some phenotypes that were overlapping. Explained later in text.)
```{r}
# extract NG_IDs and time points (N.B. time here is numeric)
mat <- MOFAobject@covariates$single_group

# convert to matrix
tib <- as_tibble(t(mat))

# set names of matrix
tib$sample <- attr(mat, "dimnames")[[2]]

# get factor scores
z_scores <- as_tibble(MOFAobject@expectations$Z$single_group)
z_scores$sample <- rownames(MOFAobject@expectations$Z$single_group)

# Phenotypes that were significant after post-hoc phenotype analysis
phenotype_data <- MOFAobject@samples_metadata[,c("NG_ID_short", "birth_deliverybirthcard_mode_binary",
                                                 "birth_deliverybirthcard_place_delivery_simple",
                                                 "infant_birthcard_feeding_mode_after_birth",
                                                 "infant_birthcard_feeding_mode_after_birth",
                                                 "mother_health_smoked_one_whole_year_p18",
                                                 "mother_birthcard_parity",
                                                 "infant_misc_sex")]

data <-
  tib %>%
  left_join(., z_scores, by = "sample") %>%
  mutate(NG_ID_short = str_sub(sample, 1, 8)) %>%
  left_join(., phenotype_data, by = c("NG_ID_short")) %>%
  mutate(across(where(is.character), ~na_if(., "<NA>"))) %>%
  mutate(birth_deliverybirthcard_mode_binary = as.factor(birth_deliverybirthcard_mode_binary))
```

## Plot Factors with Significant Phenotype Associations.
These factors had significant associations with phenotypes in association_analysis.
All trajectories are amalgamated in the same plot at the end of this code chunk.
```{r}
# fac 1 ----
#colors
col_vec <- c("#FFC20A", "#0C7BDC")
col_vec2 <- c("#FFC20A", "#0C7BDC","#004D40")
fac1_colors <- c("#fdb863", "#8073ac") # https://loading.io/color/feature/PuOr-10/ (nice selection of palettes)

fac1_trajectory_sp <-
data %>%
  filter(birth_deliverybirthcard_mode_binary != "nan") %>%
  ggplot(aes(x = time, y = Factor1, group = birth_deliverybirthcard_mode_binary, color = birth_deliverybirthcard_mode_binary)) +
  geom_point(size = 0.5) +
  scale_color_manual(values = col_vec) +
  stat_summary(
    fun = mean,   # calc mean
    geom = "line",  
    aes(group = birth_deliverybirthcard_mode_binary),  # draw lines for each group
    size = 1  
  ) +
  theme_classic() +
  xlab("time (month)") +
  ylab("factor 1 score") +
  labs(color = "birth mode")

fac1_trajectory_bp <-
data %>%
  filter(birth_deliverybirthcard_mode_binary != "nan") %>%
  ggplot(aes(x = birth_deliverybirthcard_mode_binary, y = Factor1, group = birth_deliverybirthcard_mode_binary, color = birth_deliverybirthcard_mode_binary, fill = birth_deliverybirthcard_mode_binary, alpha = 0.5)) +
  geom_boxplot(fill = col_vec) +
  scale_color_manual(values = col_vec) +
  scale_fill_manual(values = col_vec) +
  theme_void() +
  labs(fill = "birth mode",
       y = "Factor 1 Scores",
       x = "birthmode") +
  guides(color = "none") +
  scale_fill_discrete(guide="none") +
  scale_alpha(guide = 'none')

# fac 2 ----
fac2_trajectory_sp <-
  data %>%
  filter(birth_deliverybirthcard_mode_binary != "nan") %>%
  ggplot(aes(x = time, y = Factor2, group = birth_deliverybirthcard_mode_binary, color = birth_deliverybirthcard_mode_binary)) +
  geom_point(size = 0.5) +
  scale_color_manual(values = col_vec) +
  stat_summary(
    fun = mean,   
    geom = "line",  
    aes(group = birth_deliverybirthcard_mode_binary), 
    size = 1  # 
  ) +
  theme_classic() +
  xlab("time (month)") +
  ylab("factor 2 score") +
  labs(color = "birth mode")

fac2_trajectory_bp <-
  data %>%
  filter(birth_deliverybirthcard_place_delivery_simple != "nan") %>%
  ggplot(aes(x = birth_deliverybirthcard_place_delivery_simple, y = Factor2, group = birth_deliverybirthcard_place_delivery_simple, color = birth_deliverybirthcard_place_delivery_simple, fill = birth_deliverybirthcard_place_delivery_simple, alpha = 0.5)) +
  geom_boxplot(fill = fac1_colors) +
  scale_color_manual(values = col_vec) +
  scale_fill_manual(values = col_vec) +
  theme_void() +
  labs(fill = "birth place",
       y = "Factor 2 Scores",
       x = "birthplace") +
  guides(color = "none") +
  scale_fill_discrete(guide="none") +
  scale_alpha(guide = 'none')

# fac 3 ----
fac3_colors <- c("#D98481", "#91B5A9", "#7892B5")

fac3_trajectory_sp <-
data %>%
  filter(infant_birthcard_feeding_mode_after_birth != "nan") %>%
  mutate(infant_birthcard_feeding_mode_after_birth = factor(infant_birthcard_feeding_mode_after_birth, levels = c("BF", "MF", "FF") )) %>%
  ggplot(aes(x = time, y = Factor3, group = infant_birthcard_feeding_mode_after_birth, color = infant_birthcard_feeding_mode_after_birth)) +
  geom_point(size = 0.5) +
  scale_color_manual(values = col_vec2) +
  stat_summary(
    fun = mean,  
    geom = "line",  
    aes(group = infant_birthcard_feeding_mode_after_birth), 
    size = 1  
  ) +
  labs(color = "feeding mode",
      y = "Factor 3 Scores",
      x = "time (month)") +
  theme_classic() 
  
fac3_trajectory_bp <-
  data %>%
  filter(infant_birthcard_feeding_mode_after_birth != "nan") %>%
  mutate(infant_birthcard_feeding_mode_after_birth = factor(infant_birthcard_feeding_mode_after_birth, levels = c("BF", "MF", "FF") )) %>%
  ggplot(aes(x = infant_birthcard_feeding_mode_after_birth, y = Factor3, group = infant_birthcard_feeding_mode_after_birth, color = infant_birthcard_feeding_mode_after_birth, fill = infant_birthcard_feeding_mode_after_birth, alpha = 0.5)) +
  geom_boxplot(fill = col_vec2) +
  scale_color_manual(values = col_vec2) +
  #scale_fill_manual(values = fac3_colors) +
  theme_void() +
  #labs(fill = "feeding mode",
  #     y = "Factor 3 Scores",
  #     x = "birthmode") +
  guides(color = "none") +
  scale_fill_discrete(guide="none") +
  scale_alpha(guide = 'none')

# fac 4 ----
fac4_trajectory_sp <-
  data %>%
  filter(mother_health_smoked_one_whole_year_p18 != "nan") %>%
  ggplot(aes(x = time, y = Factor4, group = mother_health_smoked_one_whole_year_p18, color = mother_health_smoked_one_whole_year_p18)) +
  geom_point(size = 0.5) +
  scale_color_manual(values = fac1_colors) +
  stat_summary(
    fun = mean,  
    geom = "line",  
    aes(group = mother_health_smoked_one_whole_year_p18),  
    size = 1  # Set line size
  ) +
  theme_classic() +
  xlab("time (month)") +
  ylab("factor 4 score") +
  labs(color = "mother smoked?")

fac4_trajectory_bp <-
  data %>%
  filter(mother_health_smoked_one_whole_year_p18 != "nan") %>%
  ggplot(aes(x = mother_health_smoked_one_whole_year_p18, y = Factor4, group = mother_health_smoked_one_whole_year_p18, color = mother_health_smoked_one_whole_year_p18, fill = mother_health_smoked_one_whole_year_p18, alpha = 0.5)) +
  geom_boxplot(fill = fac1_colors) +
  scale_color_manual(values = fac1_colors) +
  scale_fill_manual(values = fac1_colors) +
  theme_void() +
  labs(fill = "smoked",
       y = "Factor 4 Scores",
       x = "birthplace") +
  guides(color = "none") +
  scale_fill_discrete(guide="none") +
  scale_alpha(guide = 'none')

# fac 5 ----
fac5_trajectory_sp <-
  data %>%
  filter(birth_deliverybirthcard_place_delivery_simple != "nan") %>%
  ggplot(aes(x = time, y = Factor5, group = birth_deliverybirthcard_place_delivery_simple, color = birth_deliverybirthcard_place_delivery_simple)) +
  geom_point(size = 0.5) +
  scale_color_manual(values = fac1_colors) +
  stat_summary(
    fun = mean,  
    geom = "line",  
    aes(group = birth_deliverybirthcard_place_delivery_simple),  
    size = 1  # Set line size
  ) +
  theme_classic() +
  xlab("time (month)") +
  ylab("factor 5 score") +
  labs(color = "delivery place")

fac5_trajectory_bp <-
  data %>%
  filter(birth_deliverybirthcard_place_delivery_simple != "nan") %>%
  ggplot(aes(x = birth_deliverybirthcard_place_delivery_simple, y = Factor4, group = birth_deliverybirthcard_place_delivery_simple, color = birth_deliverybirthcard_place_delivery_simple, fill = birth_deliverybirthcard_place_delivery_simple, alpha = 0.5)) +
  geom_boxplot(fill = fac1_colors) +
  scale_color_manual(values = fac1_colors) +
  scale_fill_manual(values = fac1_colors) +
  theme_void() +
  labs(fill = "delivery place",
       y = "Factor 5 Scores",
       x = "delivery place") +
  guides(color = "none") +
  scale_fill_discrete(guide="none") +
  scale_alpha(guide = 'none')

# fac 6 ----
fac3_colors <- c("#D98481", "#91B5A9", "#7892B5")

fac6_trajectory_sp <-
data %>%
  filter(infant_birthcard_feeding_mode_after_birth != "nan") %>%
  mutate(infant_birthcard_feeding_mode_after_birth = factor(infant_birthcard_feeding_mode_after_birth, levels = c("BF", "MF", "FF") )) %>%
  ggplot(aes(x = time, y = Factor6, group = infant_birthcard_feeding_mode_after_birth, color = infant_birthcard_feeding_mode_after_birth)) +
  geom_point(size = 0.5) +
  scale_color_manual(values = fac3_colors) +
  stat_summary(
    fun = mean,  
    geom = "line",  
    aes(group = infant_birthcard_feeding_mode_after_birth), 
    size = 1  
  ) +
  labs(color = "feeding mode",
      y = "Factor 6 Scores",
      x = "time (month)") +
  theme_classic() 
  
fac6_trajectory_bp <-
  data %>%
  filter(infant_birthcard_feeding_mode_after_birth != "nan") %>%
  mutate(infant_birthcard_feeding_mode_after_birth = factor(infant_birthcard_feeding_mode_after_birth, levels = c("BF", "MF", "FF") )) %>%
  ggplot(aes(x = infant_birthcard_feeding_mode_after_birth, y = Factor6, group = infant_birthcard_feeding_mode_after_birth, color = infant_birthcard_feeding_mode_after_birth, fill = infant_birthcard_feeding_mode_after_birth, alpha = 0.5)) +
  geom_boxplot(fill = fac3_colors) +
  scale_color_manual(values = fac3_colors) +
  #scale_fill_manual(values = fac3_colors) +
  theme_void() +
  #labs(fill = "feeding mode",
  #     y = "Factor 3 Scores",
  #     x = "birthmode") +
  guides(color = "none") +
  scale_fill_discrete(guide="none") +
  scale_alpha(guide = 'none')

# fac 8 ----
fac3_colors <- c("#D98481", "#91B5A9", "#7892B5")

fac8_trajectory_sp <-
data %>%
  filter(infant_birthcard_feeding_mode_after_birth != "nan") %>%
  mutate(infant_birthcard_feeding_mode_after_birth = factor(infant_birthcard_feeding_mode_after_birth, levels = c("BF", "MF", "FF") )) %>%
  ggplot(aes(x = time, y = Factor8, group = infant_birthcard_feeding_mode_after_birth, color = infant_birthcard_feeding_mode_after_birth)) +
  geom_point(size = 0.5) +
  scale_color_manual(values = fac3_colors) +
  stat_summary(
    fun = mean,  
    geom = "line",  
    aes(group = infant_birthcard_feeding_mode_after_birth), 
    size = 1  
  ) +
  labs(color = "feeding mode",
      y = "Factor 8 Scores",
      x = "time (month)") +
  theme_classic() 
  
fac8_trajectory_bp <-
  data %>%
  filter(infant_birthcard_feeding_mode_after_birth != "nan") %>%
  mutate(infant_birthcard_feeding_mode_after_birth = factor(infant_birthcard_feeding_mode_after_birth, levels = c("BF", "MF", "FF") )) %>%
  ggplot(aes(x = infant_birthcard_feeding_mode_after_birth, y = Factor8, group = infant_birthcard_feeding_mode_after_birth, color = infant_birthcard_feeding_mode_after_birth, fill = infant_birthcard_feeding_mode_after_birth, alpha = 0.5)) +
  geom_boxplot(fill = fac3_colors) +
  scale_color_manual(values = fac3_colors) +
  #scale_fill_manual(values = fac3_colors) +
  theme_void() +
  #labs(fill = "feeding mode",
  #     y = "Factor 3 Scores",
  #     x = "birthmode") +
  guides(color = "none") +
  scale_fill_discrete(guide="none") +
  scale_alpha(guide = 'none')

# fac 10 ----
fac10_trajectory_sp <-
  data %>%
  mutate(mother_birthcard_parity = case_when( # categorize parity
    mother_birthcard_parity == 0 ~ "First",
    mother_birthcard_parity == 1 ~ "Second",
    mother_birthcard_parity > 2 ~ "Third+",
    TRUE ~ NA
  )) %>% 
  filter(!is.na(mother_birthcard_parity)) %>%
  ggplot(aes(x = time, y = Factor10, group = mother_birthcard_parity, color = mother_birthcard_parity)) +
  geom_point(size = 0.5) +
  scale_color_manual(values = fac3_colors) +
  stat_summary(
    fun = mean,   # Calculate mean
    geom = "line",  # Use line geometry
    aes(group = mother_birthcard_parity),  
    size = 1  # Set line size
  ) +
  theme_classic() +
  xlab("time (month)") +
  ylab("factor 10 score") +
  labs(color = "pairity")

fac10_trajectory_bp <-
  data %>%
  mutate(mother_birthcard_parity = case_when( # categorize parity
    mother_birthcard_parity == 0 ~ "First",
    mother_birthcard_parity == 1 ~ "Second",
    mother_birthcard_parity >= 2 ~ "Third+",
    TRUE ~ NA
  )) %>% 
  filter(!is.na(mother_birthcard_parity)) %>%
  mutate(mother_birthcard_parity = factor(mother_birthcard_parity, levels = c("First", "Second", "Third+"))) %>%
  ggplot(aes(x = mother_birthcard_parity, y = Factor5, group = mother_birthcard_parity, color = mother_birthcard_parity, fill = mother_birthcard_parity, alpha = 0.5)) +
  geom_boxplot(fill = fac3_colors) +
  scale_color_manual(values = fac3_colors) +
  scale_fill_manual(values = fac3_colors) +
  theme_void() +
  labs(fill = "pairity",
       y = "Factor 10 Scores",
       x = "birthplace") +
  guides(color = "none") +
  scale_fill_discrete(guide="none") +
  scale_alpha(guide = 'none')


# plot all together ----
p_all_trajects <- fac1_trajectory_sp + 
                  fac2_trajectory_sp + 
                  fac3_trajectory_sp + 
                  fac4_trajectory_sp + 
                  fac5_trajectory_sp + 
                  fac6_trajectory_sp + 
                  fac8_trajectory_sp + 
                  fac10_trajectory_sp

# save plot
# set dims
indentation_mm <- 15  # Indentation in mm
height_mm <- 297 # A third of the height of an A4 page in mm (≈99 mm)

# Define the aspect ratio of your plot (width / height)
aspect_ratio <- 16/9

# Calculate the width based on the aspect ratio and adjust for indentation
width_mm <- (height_mm * aspect_ratio) - (indentation_mm * aspect_ratio)


# Save the plot as a PDF with the specified dimensions
ggsave("trajectories_16July2024.pdf", plot = p_all_trajects, width = width_mm, height = height_mm, units = "mm")
```

## Plot scatter plots of factor values
Here we are looking at our data in latent space, as in a PCA.
All individual plots appear together at end.

```{r}
## time ----
p_time <-
plot_factors(MOFAobject, color_by = "time", show_missing = FALSE) +
  labs(fill = "time",
      y = "Factor 2 Scores",
      x = "Factor 1 Scores")


## Fac1: birth mode ----
MOFAobject@samples_metadata$birth_deliverybirthcard_mode_binary <- 
  replace(MOFAobject@samples_metadata$birth_deliverybirthcard_mode_binary, MOFAobject@samples_metadata$birth_deliverybirthcard_mode_binary == "nan", NA)

p_bm <-
plot_factors(MOFAobject, factors = c(2,1), color_by = "birth_deliverybirthcard_mode_binary", show_missing = FALSE) +
  scale_fill_manual(values = fac1_colors) +
  scale_color_manual(values = fac1_colors) +
  labs(fill = "birth mode",
       y = "Factor 1 Scores",
       x = "Factor 2 Scores")

## Fac2: birth place ----
MOFAobject@samples_metadata$birth_deliverybirthcard_place_delivery_simple <- 
  replace(MOFAobject@samples_metadata$birth_deliverybirthcard_place_delivery_simple, MOFAobject@samples_metadata$birth_deliverybirthcard_place_delivery_simple == "nan", NA)

p_bp <-
plot_factors(MOFAobject, factors = c(1,2), color_by = "birth_deliverybirthcard_mode_binary", show_missing = FALSE) +
  scale_fill_manual(values = fac1_colors) +
  scale_color_manual(values = fac1_colors) +
  labs(fill = "birth place",
       y = "Factor 2 Scores",
       x = "Factor 1 Scores")

## Fac3: feeding mode ----
MOFAobject@samples_metadata$infant_birthcard_feeding_mode_after_birth <- 
  replace(MOFAobject@samples_metadata$infant_birthcard_feeding_mode_after_birth, MOFAobject@samples_metadata$infant_birthcard_feeding_mode_after_birth == "nan", NA)

MOFAobject@samples_metadata$infant_birthcard_feeding_mode_after_birth <- 
  factor(MOFAobject@samples_metadata$infant_birthcard_feeding_mode_after_birth, 
         levels = c("BF", "MF", "FF", NA))

p_fm <-
  plot_factors(MOFAobject, factors = c(1,3), color_by = "infant_birthcard_feeding_mode_after_birth", show_missing = FALSE) +
  scale_fill_manual(values = fac3_colors) +
  scale_color_manual(values = fac3_colors) +
  labs(fill = "feeding mode",
       y = "Factor 3 Scores",
       x = "Factor 1 Scores")

## Fac4: smoking ----
MOFAobject@samples_metadata$mother_health_smoked_one_whole_year_p18 <- 
  replace(MOFAobject@samples_metadata$mother_health_smoked_one_whole_year_p18, MOFAobject@samples_metadata$mother_health_smoked_one_whole_year_p18 == "nan", NA)

p_smk <-
  plot_factors(MOFAobject, factors = c(1,4), color_by = "mother_health_smoked_one_whole_year_p18", show_missing = FALSE) +
  scale_fill_manual(values = fac1_colors) +
  scale_color_manual(values = fac1_colors) +
  labs(fill = "smoked?",
       y = "Factor 4 Scores",
       x = "Factor 1 Scores")

## Fac5: delivery place ----
MOFAobject@samples_metadata$birth_deliverybirthcard_place_delivery_simple <- 
  replace(MOFAobject@samples_metadata$birth_deliverybirthcard_place_delivery_simple, MOFAobject@samples_metadata$birth_deliverybirthcard_place_delivery_simple == "nan", NA)

p_dp <-
  plot_factors(MOFAobject, factors = c(1,5), color_by = "birth_deliverybirthcard_place_delivery_simple", show_missing = FALSE) +
  scale_fill_manual(values = fac1_colors) +
  scale_color_manual(values = fac1_colors) +
  labs(fill = "delivery place",
       y = "Factor 5 Scores",
       x = "Factor 1 Scores")

## Fac6: delivery place ----
MOFAobject@samples_metadata$infant_birthcard_feeding_mode_after_birth <- 
  replace(MOFAobject@samples_metadata$infant_birthcard_feeding_mode_after_birth, MOFAobject@samples_metadata$infant_birthcard_feeding_mode_after_birth == "nan", NA)

p_fm_f6 <-
  plot_factors(MOFAobject, factors = c(1,6), color_by = "infant_birthcard_feeding_mode_after_birth", show_missing = FALSE) +
  scale_fill_manual(values = fac3_colors) +
  scale_color_manual(values = fac3_colors) +
  labs(fill = "feeding mode",
       y = "Factor 6 Scores",
       x = "Factor 1 Scores")

## Fac8: feeding mode----
MOFAobject@samples_metadata$infant_birthcard_feeding_mode_after_birth <- 
  replace(MOFAobject@samples_metadata$infant_birthcard_feeding_mode_after_birth, MOFAobject@samples_metadata$infant_birthcard_feeding_mode_after_birth == "nan", NA)

p_fm_f8 <-
  plot_factors(MOFAobject, factors = c(1,8), color_by = "infant_birthcard_feeding_mode_after_birth", show_missing = FALSE) +
  scale_fill_manual(values = fac3_colors) +
  scale_color_manual(values = fac3_colors) +
  labs(fill = "feeding mode",
       y = "Factor 8 Scores",
       x = "Factor 1 Scores")

## Fac10: parity ----
# Replace values in the vector directly
MOFAobject@samples_metadata$mother_birthcard_parity <- 
  ifelse(MOFAobject@samples_metadata$mother_birthcard_parity == 0, "First",
         ifelse(MOFAobject@samples_metadata$mother_birthcard_parity == 1, "Second",
                ifelse(MOFAobject@samples_metadata$mother_birthcard_parity >= 2, "Third+", NA)))

p_par <-
  plot_factors(MOFAobject, factors = c(1,10), color_by = "mother_birthcard_parity", show_missing = FALSE) +
  scale_fill_manual(values = fac3_colors) +
  scale_color_manual(values = fac3_colors) +
  labs(fill = "pairity",
       y = "Factor 10 Scores",
       x = "Factor 1 Scores")


# plot trajectories as box plots ----

########
#
design <-
  "
AC
BD
"
p1 <-
  (guide_area()) + 
  p_bm + 
  guide_area() + 
  fac1_trajectory_bp +
  plot_layout(
    widths = c(1, 0.125),
    heights = c(0.125, 1),
    design = design
  ) 

p2 <-
  (guide_area()) + 
  p_bp + 
  guide_area() + 
  fac2_trajectory_bp +
  plot_layout(
    widths = c(1, 0.125),
    heights = c(0.125, 1),
    design = design
  )

p3 <-
  (guide_area()) + 
  p_fm + 
  guide_area() + 
  fac3_trajectory_bp +
  plot_layout(
    widths = c(1, 0.125),
    heights = c(0.125, 1),
    design = design
  )

p4 <-
  (guide_area()) + 
  p_smk + 
  guide_area() + 
  fac4_trajectory_bp +
  plot_layout(
    widths = c(1, 0.125),
    heights = c(0.125, 1),
    design = design
  )

p5 <-
  (guide_area()) + 
  p_dp + 
  guide_area() + 
  fac5_trajectory_bp +
  plot_layout(
    widths = c(1, 0.125),
    heights = c(0.125, 1),
    design = design
  )

p6 <-
  (guide_area()) + 
  p_fm_f6 + 
  guide_area() + 
  fac6_trajectory_bp +
  plot_layout(
widths = c(1, 0.125),
    heights = c(0.125, 1),
    design = design
  )

p8 <-
  (guide_area()) + 
  p_fm_f8 + 
  guide_area() + 
  fac8_trajectory_bp +
  plot_layout(
widths = c(1, 0.125),
    heights = c(0.125, 1),
    design = design
  )

p10 <-
  (guide_area()) + 
  p_par + 
  guide_area() + 
  fac10_trajectory_bp +
  plot_layout(
widths = c(1, 0.125),
    heights = c(0.125, 1),
    design = design
  )

p11 <-
  (guide_area()) + 
  p_time + 
  guide_area() + 
  guide_area() +
  plot_layout(
    widths = c(1, 0.125),
    heights = c(0.125, 1),
    design = design
  ) 

p_scatters <- (p1 | p2 | p3) / (p4 | p5 | p6) / (p8 | p10 | p11) # with patchwork library

# save plot
# set dims
indentation_mm <- 15  # in mm
height_mm <- 297 

# define the aspect ratio 
aspect_ratio <- 16/9

# calc width
width_mm <- (height_mm * aspect_ratio) - (indentation_mm * aspect_ratio)


# save as .pdf
ggsave("scatterplots_16July2024.pdf", plot = p_scatters, width = width_mm, height = height_mm, units = "mm")
```

## plot weights 
These are the species that contribute the most to variation explained across each factor.

```{r, warning=FALSE}
## collapse to species ----
feature_matrix <-
  feature_matrix %>% rename(feature_description = taxon, feature = taxonID)

features_metadata(MOFAobject) <- features_metadata(MOFAobject) %>%
  left_join(.,feature_matrix, by = c("feature")) %>%
  separate(col = feature_description, sep = "\\.",
           into = c("kingdom", "phylum", "class",
                    "order", "family", "genus", "species","taxon"),
           remove = FALSE)

# Fac 1 weights----
col_vec <- c("#FFC20A", "#0C7BDC")

df_weights1 <- get_weights(MOFAobject, as.data.frame = T, factors = 1) %>%
  left_join(., features_metadata(MOFAobject), by = c("feature", "view"))

# filter out genera with missing genera information and shorten name
df_weights1 %<>%
  filter(is.na(taxon) | !str_detect(taxon, "^t__")) %>% 
  select(-taxon) %>% 
  filter(!is.na(genus)) %>% 
  filter(!is.na(species)) %>%
  mutate(species = gsub("s__","", species)) 


# summarize by mean weights across all species in the genus and 
# filter to top 10 positive and negative ones
df_top <- df_weights1 %>% group_by(species) %>% select(-feature_description) %>% 
  summarize(mean_weight = mean(value), n_spec = n()) %>% 
  ungroup() %>%
  # cesarean is linked to higher latent values
  mutate(type = ifelse(mean_weight < 0, "Negative Weights", "Positive Weights")) %>%
  group_by(type) %>%
  slice_max(abs(mean_weight), n= 5) %>% ungroup() %>% arrange(mean_weight) %>%
  mutate(species = factor(species, levels = .$species))

fac1_weights <-
  df_top %>% 
  mutate(birthmode = case_when(
    type == "Negative Weights" ~ "CS",
    type == "Positive Weights" ~ "VG",
    TRUE ~ "undetermined"  
  )) %>%
  ggplot(aes(x= species, y = mean_weight, fill = birthmode)) + 
  geom_bar(stat="identity") +
  coord_flip() + 
  #theme_bw()  + 
  #theme(legend.position = "top") + 
  xlab("") + 
  #guides(fill = guide_legend(title="")) +
  ggtitle("Influence of Species Across Factor") +
  geom_point(data = filter(df_weights1, species %in% df_top$species),
             aes(x = species, y = value), inherit.aes = FALSE, alpha = 0.3)  +
  scale_fill_manual(values = col_vec) +
  ylab("Weight (Factor 1)") +
  theme_classic() +
  theme(#plot.title = element_text(hjust = 0.5),
    text = element_text(size = 10),
    plot.title = element_blank())

fac1_weights

# Fac 2 weights----
df_weights1 <- get_weights(MOFAobject, as.data.frame = T, factors = 2) %>%
  left_join(., features_metadata(MOFAobject), by = c("feature", "view"))

# filter out genera with missing genera information and shorten name
df_weights1 %<>%
  filter(is.na(taxon) | !str_detect(taxon, "^t__")) %>% 
  select(-taxon) %>% 
  filter(!is.na(genus)) %>% 
  filter(!is.na(species)) %>%
  mutate(species = gsub("s__","", species)) 


# summarize by mean weights across all species in the genus and 
# filter to top 10 positive and negative ones
df_top <- df_weights1 %>% group_by(species) %>% select(-feature_description) %>% 
  summarize(mean_weight = mean(value), n_spec = n()) %>% 
  ungroup() %>%
  # cesarean is linked to higher latent values
  mutate(type = ifelse(mean_weight < 0, "Negative Weights", "Positive Weights")) %>%
  group_by(type) %>%
  slice_max(abs(mean_weight), n= 5) %>% ungroup() %>% arrange(mean_weight) %>%
  mutate(species = factor(species, levels = .$species))

fac2_weights <-
  df_top %>% 
  mutate(birthmode = case_when(
    type == "Negative Weights" ~ "VG",
    type == "Positive Weights" ~ "CS",
    TRUE ~ "undetermined"  
  )) %>%
  ggplot(aes(x= species, y = mean_weight, fill = birthmode)) + 
  geom_bar(stat="identity") +
  coord_flip() + 
  #theme_bw()  + 
  #theme(legend.position = "top") + 
  xlab("") + 
  #guides(fill = guide_legend(title="")) +
  ggtitle("Influence of Species Across Factor") +
  geom_point(data = filter(df_weights1, species %in% df_top$species),
             aes(x = species, y = value), inherit.aes = FALSE, alpha = 0.3)  +
  scale_fill_manual(values = col_vec) +
  ylab("Weight (Factor 2)") +
  theme_classic() +
  theme(#plot.title = element_text(hjust = 0.5),
    text = element_text(size = 10),
    plot.title = element_blank())

fac2_weights

# fac3 weights ----
col_vec3 <- c("#FFC20A","#004D40")

df_weights1 <- get_weights(MOFAobject, as.data.frame = T, factors = 3) %>%
  left_join(., features_metadata(MOFAobject), by = c("feature", "view"))

# filter out genera with missing genera information and shorten name
df_weights1 %<>%
  filter(is.na(taxon) | !str_detect(taxon, "^t__")) %>% 
  select(-taxon) %>% 
  filter(!is.na(genus)) %>% 
  filter(!is.na(species)) %>%
  mutate(species = gsub("s__","", species)) 


# summarize by mean weights across all species in the genus and 
# filter to top 10 positive and negative ones
df_top <- df_weights1 %>% group_by(species) %>% select(-feature_description) %>% 
  summarize(mean_weight = mean(value), n_spec = n()) %>% 
  ungroup() %>%
  # cesarean is linked to higher latent values
  mutate(type = ifelse(mean_weight < 0, "Negative Weights", "Positive Weights")) %>%
  group_by(type) %>%
  slice_max(abs(mean_weight), n= 5) %>% ungroup() %>% arrange(mean_weight) %>%
  mutate(species = factor(species, levels = .$species))

fac3_weights <-
  df_top %>% 
  mutate(feedingmode = case_when(
    type == "Negative Weights" ~ "FF",
    type == "Positive Weights" ~ "BF",
    TRUE ~ "undetermined"  
  )) %>%
  ggplot(aes(x= species, y = mean_weight, fill = feedingmode)) + 
  geom_bar(stat="identity") +
  coord_flip() + 
  #theme_bw()  + 
  #theme(legend.position = "top") + 
  xlab("") + 
  #guides(fill = guide_legend(title="")) +
  ggtitle("Influence of Species Across Factor") +
  geom_point(data = filter(df_weights1, species %in% df_top$species),
             aes(x = species, y = value), inherit.aes = FALSE, alpha = 0.3)  +
  scale_fill_manual(values = col_vec3) +
  ylab("Weight (Factor 3)") +
  theme_classic() +
  theme(#plot.title = element_text(hjust = 0.5),
    text = element_text(size = 10),
    plot.title = element_blank())

fac3_weights

# fac4 weights ----
fac4_colors_weights  <- c("#fdb863", "#8073ac")


df_weights1 <- get_weights(MOFAobject, as.data.frame = T, factors = 4) %>%
  left_join(., features_metadata(MOFAobject), by = c("feature", "view"))

# filter out genera with missing genera information and shorten name
df_weights1 %<>%
  filter(is.na(taxon) | !str_detect(taxon, "^t__")) %>% 
  select(-taxon) %>% 
  filter(!is.na(genus)) %>% 
  filter(!is.na(species)) %>%
  mutate(species = gsub("s__","", species)) 


# summarize by mean weights across all species in the genus and 
# filter to top 10 positive and negative ones
df_top <- df_weights1 %>% group_by(species) %>% select(-feature_description) %>% 
  summarize(mean_weight = mean(value), n_spec = n()) %>% 
  ungroup() %>%
  # cesarean is linked to higher latent values
  mutate(type = ifelse(mean_weight < 0, "Negative Weights", "Positive Weights")) %>%
  group_by(type) %>%
  slice_max(abs(mean_weight), n= 5) %>% ungroup() %>% arrange(mean_weight) %>%
  mutate(species = factor(species, levels = .$species))

fac4_weights <-
  df_top %>% 
  mutate(smoking = case_when(
    type == "Negative Weights" ~ "yes",
    type == "Positive Weights" ~ "no",
    TRUE ~ "undetermined"  
  )) %>%
  ggplot(aes(x= species, y = mean_weight, fill = smoking)) + 
  geom_bar(stat="identity") +
  coord_flip() + 
  #theme_bw()  + 
  #theme(legend.position = "top") + 
  xlab("") + 
  #guides(fill = guide_legend(title="")) +
  ggtitle("Influence of Species Across Factor") +
  geom_point(data = filter(df_weights1, species %in% df_top$species),
             aes(x = species, y = value), inherit.aes = FALSE, alpha = 0.3)  +
  scale_fill_manual(values = fac4_colors_weights) +
  ylab("Weight (Factor 4)") +
  theme_classic() +
  theme(#plot.title = element_text(hjust = 0.5),
    text = element_text(size = 10),
    plot.title = element_blank())

fac4_weights

# fac5 weights ----
fac5_colors_weights  <- c("#fdb863", "#8073ac")

df_weights1 <- get_weights(MOFAobject, as.data.frame = T, factors = 5) %>%
  left_join(., features_metadata(MOFAobject), by = c("feature", "view"))

# filter out genera with missing genera information and shorten name
df_weights1 %<>%
  filter(is.na(taxon) | !str_detect(taxon, "^t__")) %>% 
  select(-taxon) %>% 
  filter(!is.na(genus)) %>% 
  filter(!is.na(species)) %>%
  mutate(species = gsub("s__","", species)) 


# summarize by mean weights across all species in the genus and 
# filter to top 10 positive and negative ones
df_top <- df_weights1 %>% group_by(species) %>% select(-feature_description) %>% 
  summarize(mean_weight = mean(value), n_spec = n()) %>% 
  ungroup() %>%
  # cesarean is linked to higher latent values
  mutate(type = ifelse(mean_weight < 0, "Negative Weights", "Positive Weights")) %>%
  group_by(type) %>%
  slice_max(abs(mean_weight), n= 5) %>% ungroup() %>% arrange(mean_weight) %>%
  mutate(species = factor(species, levels = .$species))

fac5_weights <-
  df_top %>% 
  mutate(deliveryplace = case_when(
    type == "Negative Weights" ~ "hospital",
    type == "Positive Weights" ~ "home",
    TRUE ~ "undetermined"  
  )) %>%
  ggplot(aes(x= species, y = mean_weight, fill = deliveryplace)) + 
  geom_bar(stat="identity") +
  coord_flip() + 
  #theme_bw()  + 
  #theme(legend.position = "top") + 
  xlab("") + 
  #guides(fill = guide_legend(title="")) +
  ggtitle("Influence of Species Across Factor") +
  geom_point(data = filter(df_weights1, species %in% df_top$species),
             aes(x = species, y = value), inherit.aes = FALSE, alpha = 0.3)  +
  scale_fill_manual(values = col_vec) +
  ylab("Weight (Factor 5)") +
  theme_classic() +
  theme(#plot.title = element_text(hjust = 0.5),
    text = element_text(size = 10),
    plot.title = element_blank())

fac5_weights

# fac6 weights ----
df_weights1 <- get_weights(MOFAobject, as.data.frame = T, factors = 6) %>%
  left_join(., features_metadata(MOFAobject), by = c("feature", "view"))

# filter out genera with missing genera information and shorten name
df_weights1 %<>%
  filter(is.na(taxon) | !str_detect(taxon, "^t__")) %>% 
  select(-taxon) %>% 
  filter(!is.na(genus)) %>% 
  filter(!is.na(species)) %>%
  mutate(species = gsub("s__","", species)) 


# summarize by mean weights across all species in the genus and 
# filter to top 10 positive and negative ones
df_top <- df_weights1 %>% group_by(species) %>% select(-feature_description) %>% 
  summarize(mean_weight = mean(value), n_spec = n()) %>% 
  ungroup() %>%
  # cesarean is linked to higher latent values
  mutate(type = ifelse(mean_weight < 0, "Negative Weights", "Positive Weights")) %>%
  group_by(type) %>%
  slice_max(abs(mean_weight), n= 5) %>% ungroup() %>% arrange(mean_weight) %>%
  mutate(species = factor(species, levels = .$species))

fac6_weights <-
  df_top %>% 
  mutate(feedingmode = case_when(
    type == "Negative Weights" ~ "MF",
    type == "Positive Weights" ~ "FF",
    TRUE ~ "undetermined"  
  )) %>%
  ggplot(aes(x= species, y = mean_weight, fill = feedingmode)) + 
  geom_bar(stat="identity") +
  coord_flip() + 
  #theme_bw()  + 
  #theme(legend.position = "top") + 
  xlab("") + 
  #guides(fill = guide_legend(title="")) +
  ggtitle("Influence of Species Across Factor") +
  geom_point(data = filter(df_weights1, species %in% df_top$species),
             aes(x = species, y = value), inherit.aes = FALSE, alpha = 0.3)  +
  scale_fill_manual(values = col_vec) +
  ylab("Weight (Factor 6)") +
  theme_classic() +
  theme(#plot.title = element_text(hjust = 0.5),
    text = element_text(size = 10),
    plot.title = element_blank())

fac6_weights

# fac8 weights ----
df_weights1 <- get_weights(MOFAobject, as.data.frame = T, factors = 8) %>%
  left_join(., features_metadata(MOFAobject), by = c("feature", "view"))

# filter out genera with missing genera information and shorten name
df_weights1 %<>%
  filter(is.na(taxon) | !str_detect(taxon, "^t__")) %>% 
  select(-taxon) %>% 
  filter(!is.na(genus)) %>% 
  filter(!is.na(species)) %>%
  mutate(species = gsub("s__","", species)) 


# summarize by mean weights across all species in the genus and 
# filter to top 10 positive and negative ones
df_top <- df_weights1 %>% group_by(species) %>% select(-feature_description) %>% 
  summarize(mean_weight = mean(value), n_spec = n()) %>% 
  ungroup() %>%
  # cesarean is linked to higher latent values
  mutate(type = ifelse(mean_weight < 0, "Negative Weights", "Positive Weights")) %>%
  group_by(type) %>%
  slice_max(abs(mean_weight), n= 5) %>% ungroup() %>% arrange(mean_weight) %>%
  mutate(species = factor(species, levels = .$species))

fac8_weights <-
  df_top %>% 
  mutate(feedingmode = case_when(
    type == "Negative Weights" ~ "BF",
    type == "Positive Weights" ~ "FF",
    TRUE ~ "undetermined"  
  )) %>%
  ggplot(aes(x= species, y = mean_weight, fill = feedingmode)) + 
  geom_bar(stat="identity") +
  coord_flip() + 
  #theme_bw()  + 
  #theme(legend.position = "top") + 
  xlab("") + 
  #guides(fill = guide_legend(title="")) +
  ggtitle("Influence of Species Across Factor") +
  geom_point(data = filter(df_weights1, species %in% df_top$species),
             aes(x = species, y = value), inherit.aes = FALSE, alpha = 0.3)  +
  scale_fill_manual(values = col_vec) +
  ylab("Weight (Factor 8)") +
  theme_classic() +
  theme(#plot.title = element_text(hjust = 0.5),
    text = element_text(size = 10),
    plot.title = element_blank())

fac8_weights

# fac10 weights ----
df_weights1 <- get_weights(MOFAobject, as.data.frame = T, factors = 10) %>%
  left_join(., features_metadata(MOFAobject), by = c("feature", "view"))

# filter out genera with missing genera information and shorten name
df_weights1 %<>%
  filter(is.na(taxon) | !str_detect(taxon, "^t__")) %>% 
  select(-taxon) %>% 
  filter(!is.na(genus)) %>% 
  filter(!is.na(species)) %>%
  mutate(species = gsub("s__","", species)) 


# summarize by mean weights across all species in the genus and 
# filter to top 10 positive and negative ones
df_top <- df_weights1 %>% group_by(species) %>% select(-feature_description) %>% 
  summarize(mean_weight = mean(value), n_spec = n()) %>% 
  ungroup() %>%
  # cesarean is linked to higher latent values
  mutate(type = ifelse(mean_weight < 0, "Negative Weights", "Positive Weights")) %>%
  group_by(type) %>%
  slice_max(abs(mean_weight), n= 5) %>% ungroup() %>% arrange(mean_weight) %>%
  mutate(species = factor(species, levels = .$species))

fac10_weights <-
  df_top %>% 
  mutate(parity = case_when(
    type == "Negative Weights" ~ "Third+",
    type == "Positive Weights" ~ "First",
    TRUE ~ "undetermined"  
  )) %>%
  ggplot(aes(x= species, y = mean_weight, fill = parity)) + 
  geom_bar(stat="identity") +
  coord_flip() + 
  #theme_bw()  + 
  #theme(legend.position = "top") + 
  xlab("") + 
  #guides(fill = guide_legend(title="")) +
  ggtitle("Influence of Species Across Factor") +
  geom_point(data = filter(df_weights1, species %in% df_top$species),
             aes(x = species, y = value), inherit.aes = FALSE, alpha = 0.3)  +
  scale_fill_manual(values = col_vec) +
  ylab("Weight (Factor 10)") +
  theme_classic() +
  theme(#plot.title = element_text(hjust = 0.5),
    text = element_text(size = 10),
    plot.title = element_blank())

fac10_weights


# plot all wieghts
p_weights <- fac1_weights + fac2_weights + fac3_weights + fac4_weights + fac5_weights + fac6_weights + fac8_weights + fac10_weights

# plots wieghts factors 1 through 3
p_weights_v2 <- fac1_weights + fac2_weights + fac3_weights

# save plot
ggsave("weights.png", plot = p_weights_v2, width = 15, height = 5)
```


## Plot Factor One Through Three as Boxplots
```{r}
# fac 1 ----
col_vec <- c("#FFC20A", "#0C7BDC")
col_vec2 <- c("#FFC20A", "#0C7BDC","#004D40")

fac1_trajectory_sp <-
data %>% 
  filter(birth_deliverybirthcard_mode_binary != "nan") %>%
  mutate(
    time = factor(time, levels = c(0.5, 1, 2, 3, 6, 9, 12)) 
  ) %>%
  ggplot(aes(x = time, y = Factor1, group = interaction(time, birth_deliverybirthcard_mode_binary), color = birth_deliverybirthcard_mode_binary, fill = birth_deliverybirthcard_mode_binary)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.7), size = 0.3) +
  geom_boxplot(alpha = 0.4, position = position_dodge(width = 0.7), width = 0.6,outlier.shape = NA) +
  scale_color_manual(values = col_vec) +
  scale_fill_manual(values = col_vec) +
  stat_summary(
    fun = mean,   # calc mean
    geom = "line",  
    aes(group = birth_deliverybirthcard_mode_binary),  # draw lines for each group
    size = 0.75,
    alpha = 0.4
  ) +
  scale_x_discrete(
    labels = c("W2", "M1", "M2", "M3", "M6", "M9", "M12")  
  ) +
  theme_classic() +
  xlab("time (month)") +
  ylab("factor 1 score") +
  labs(fill = "birth mode") +
  guides(color = "none")  

# Assuming your plot is stored in a variable called p
saveRDS(fac1_trajectory_sp, file = "fac1_trajectory_sp.rds")


# fac 2 ----
fac2_trajectory_sp <-
data %>% 
  filter(birth_deliverybirthcard_mode_binary != "nan") %>%
  mutate(
    time = factor(time, levels = c(0.5, 1, 2, 3, 6, 9, 12))  
  ) %>%
  ggplot(aes(x = time, y = Factor2, group = interaction(time, birth_deliverybirthcard_mode_binary), color = birth_deliverybirthcard_mode_binary, fill = birth_deliverybirthcard_mode_binary)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.7), size = 0.3) +
  geom_boxplot(alpha = 0.4, position = position_dodge(width = 0.7), width = 0.6,outlier.shape = NA) +
  scale_color_manual(values = col_vec) +
  scale_fill_manual(values = col_vec) +
  stat_summary(
    fun = mean,   # calc mean
    geom = "line",  
    aes(group = birth_deliverybirthcard_mode_binary),  # draw lines for each group
    size = 0.75,
    alpha = 0.4
  ) +
  scale_x_discrete(
    labels = c("W2", "M1", "M2", "M3", "M6", "M9", "M12")  
  ) +
  theme_classic() +
  xlab("time (month)") +
  ylab("factor 2 score") +
  labs(fill = "birth mode") +
  guides(color = "none")  

saveRDS(fac2_trajectory_sp, file = "fac2_trajectory_sp.rds")

# fac 3 ----
fac3_trajectory_sp <-
data %>%
  filter(infant_birthcard_feeding_mode_after_birth != "nan") %>%
  mutate(
    infant_birthcard_feeding_mode_after_birth = factor(infant_birthcard_feeding_mode_after_birth, levels = c("BF", "MF", "FF")),
    time = factor(time, levels = c(0.5, 1, 2, 3, 6, 9, 12)) 
  ) %>%
  mutate(infant_birthcard_feeding_mode_after_birth = factor(infant_birthcard_feeding_mode_after_birth, levels = c("BF", "MF", "FF") )) %>%
  ggplot(aes(x = time, y = Factor3, group = interaction(time, infant_birthcard_feeding_mode_after_birth), color = infant_birthcard_feeding_mode_after_birth, fill  = infant_birthcard_feeding_mode_after_birth)) +
  geom_point(position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.7), size = 0.3) +
  geom_boxplot(alpha = 0.4, position = position_dodge(width = 0.7), width = 0.6, outlier.shape = NA) +
  scale_color_manual(values = col_vec2) +
  scale_fill_manual(values = col_vec2) +
  stat_summary(
    fun = mean,
    geom = "line",
    aes(group = infant_birthcard_feeding_mode_after_birth),
    size = 0.75,
    alpha = 0.4
  ) +
  scale_x_discrete(
    labels = c("W2", "M1", "M2", "M3", "M6", "M9", "M12")  
  ) +
  labs(fill = "feeding mode",
       y = "Factor 3 Scores",
       x = "time (month)") +
  theme_classic() +
  guides(color = "none")  

saveRDS(fac3_trajectory_sp, file = "fac3_trajectory_sp.rds")


trajectories_1_to_3 <- fac1_trajectory_sp + fac2_trajectory_sp + fac3_trajectory_sp

# save plot
ggsave("trajectories_1_to_3.png", trajectories_1_to_3, width = 15 , height = 5 , dpi = 600)

fac1_trajectory_sp <- readRDS("fac1_trajectory_sp.rds")
fac2_trajectory_sp <- readRDS("fac2_trajectory_sp.rds")
fac3_trajectory_sp <- readRDS("fac3_trajectory_sp.rds")
```


# Mixed Modelling: factors ~ phenotypes
The purpose of this analysis is to associate latent variables (i.e, factors) with phenotypes. This was employed in a mixed modelling strategy where factor scores were associated with high priority phenotypes from LL-NEXT. I did remove three phenotypes (birth_delivery_mode_complex, birth_delivery_mode_simple, infant_ffq_ever_never_breastfed) because they  are likely co-correlated with other phenotypes, namely delivery_mode_binary and infant_birthcard_feeding_mode_after_birth. 


The modelling strategy I employed can be described as a _sequential association analysis with a stepwise correction_. Such a strategy aligns well with a _fine mapping_ strategy used in genetics to identify genetic variants associated with phenotypes. This strategy involved running mixed models for phenotypes individually, while adjusting for covariates and time. Then, after fitting these initial models, the phenotype with the lowest p-value (i.e., strongest association) is identified and used as an additional covariate in another round of associations with all other phenotypes. 

The goal is to control for confounding effects driven by the strongest phenotype association. This allows more precise estimations of variance and p-value calculations that include less strongly associated but nonetheless important phenotypes that drive variation across factors. We expect that if the adjusted p-value of a corrected phenotype is lower that the adjusted p-value of the most strongly associated phenotype in the first round, then we would have identified a model that contains more precise variance and pvalue estimations. Furthermore, the corrected phenotype in question would then replace the phenotype found to be the strongest association in the first round. Although this strategy does not correct for most strongly associated phenotype _per se_, it does allow one to tease apart independent effects of less strongly associated phenotypes on factor scores.

Except on factor 10, the execution of this strategy did not alter the most strongly associated phenotype.  


## Clear Env and Load Libraries and Data
```{r, warning=FALSE, message = FALSE}
# clear environment
rm(list = ls())

# load libraries
library(kableExtra)
library(lme4)
library(lmerTest)
library(broom)
library(broom.mixed)
library(tidyverse)
library(MOFA2)
library(here)
library(openxlsx)

# load data used for MEFISTO ----
load(here("data_for_MOFA_26_01_2024.rda"))
head(data_for_MOFA)

# load data phenotypes
source(here("prep_phenotypes.R")) # script that prep phenotypes
data_phenos <- 
  phenos %>%
  rename(sample = NG_ID)
head(data_phenos)

# load MEFISTO results ----
MOFAobject <- load_model(here("LLNEXT_microbiome_model_26_01_2024.hdf5"), load_interpol_Z = FALSE)
```

## Tidy Data for Modelling
```{r}
# make vector of phenotypes---
phenotypes_and_covariates <- names(data_phenos)[c(3:19)]
phenotypes_to_test <- phenotypes_and_covariates[5:17]

data_factors <-
  get_factors(MOFAobject, as.data.frame = TRUE, factors = c(1:10)) 

data_to_test <-
  data_phenos %>% 
  left_join(., data_factors, by = "sample") %>%
  ungroup()
```


## Model Function
Combined function to run mixed models. An initial set of associations is run. Then a second set of associations is run, correcting for the most significant phenotype in the initial set of associations. 
```{r}
# function to run mixed models with stepwise correction 
run_and_correct_MM <- function(df, phenotypes, factor_name) {
  initial_results <- list()
  corrected_results <- list()
  
  # run initial mixed models
  for (i in seq_along(phenotypes)) {
    phenotype <- phenotypes[i]
    
    # remove NAs from data and select needed data 
    df_filtered <- df %>%
      ungroup() %>%
      filter(factor == factor_name) %>%
      select(value, clean_reads_FQ_1, dna_conc, BATCH_NUMBER, all_of(phenotype), Timepoint_categorical, NEXT_ID) %>%
      mutate(Timepoint_categorical = factor(Timepoint_categorical, levels = c("W2", "M1", "M2","M3","M6","M9","M12"))) %>%
      filter(if_all(everything(), ~ !is.na(.)))
    
    # fit model with phenotype
    fit1 <- tryCatch({
      model <- lmerTest::lmer(as.formula(paste("value ~", "clean_reads_FQ_1 + dna_conc + BATCH_NUMBER +", phenotype, "* Timepoint_categorical + (1|NEXT_ID)")), data = df_filtered, REML = FALSE)
      list(success=TRUE, model=model, summary=broom::glance(model), tidied=broom::tidy(model) %>% mutate(phenotype = phenotype, factor = factor_name))
    }, error=function(e) {
      list(success=FALSE, error_message=conditionMessage(e), model=NULL, summary=NULL, tidied=NULL)
    }, warning = function(w) {
      list(success=FALSE, warning_message=conditionMessage(w), model=NULL, summary=NULL, tidied=NULL)
    })
    
    # fit model without phenotype
    fit0 <- tryCatch({
      model <- lmerTest::lmer(as.formula(paste("value ~", "clean_reads_FQ_1 + dna_conc + BATCH_NUMBER + Timepoint_categorical + (1|NEXT_ID)")), data = df_filtered, REML = FALSE)
      list(success=TRUE, model=model, summary=broom::glance(model), tidied=broom::tidy(model) %>% mutate(phenotype = phenotype, factor = factor_name))
    }, error=function(e) {
      list(success=FALSE, error_message=conditionMessage(e), model=NULL, summary=NULL, tidied=NULL)
    }, warning = function(w) {
      list(success=FALSE, warning_message=conditionMessage(w), model=NULL, summary=NULL, tidied=NULL)
    })
    
    # compare models if both were successfully fitted
    if (fit1$success && fit0$success) {
      model_comparison <- anova(fit1$model, fit0$model, test = "LRT")
      model_comparison_tidied <- broom::tidy(model_comparison) %>% mutate(phenotype = phenotype, factor = factor_name)
    } else {
      model_comparison <- NULL
      model_comparison_tidied <- NULL
    }
    
    # store all results 
    initial_results[[i]] <- list(fit1=fit1, fit0=fit0, model_comparison=model_comparison, model_comparison_tidied=model_comparison_tidied)
  }
  
  # summarize the initial model comparisons 
  summary_df <- data.frame()
  for (i in seq_along(initial_results)) {
    model_comparison_tidied <- initial_results[[i]]$model_comparison_tidied
    
    if (!is.null(model_comparison_tidied)) {
      summary_df <- bind_rows(summary_df, model_comparison_tidied) %>%  filter(!is.na(p.value)) %>%  mutate(p.adj = p.adjust(p.value, method = "bonferroni")) 
    }
  }
  
  # identify the phenotype with the lowest p-value
  lowest_p_phenotype <- summary_df %>% filter(!is.na(p.value)) %>% arrange(p.adj) %>% slice(1) %>% pull(phenotype)
  
  # corrected Associations: correct other phenotypes using the phenotype with the lowest p-value
  for (phenotype in phenotypes) {
    if (phenotype != lowest_p_phenotype) {
      # Remove NAs from data
      df_filtered <- df %>%
        ungroup() %>%
        filter(factor == factor_name) %>%
        select(value, clean_reads_FQ_1, dna_conc, BATCH_NUMBER, all_of(lowest_p_phenotype), all_of(phenotype), Timepoint_categorical, NEXT_ID) %>%
        mutate(Timepoint_categorical = factor(Timepoint_categorical, levels = c("W2", "M1", "M2","M3","M6","M9","M12"))) %>%
        filter(if_all(everything(), ~ !is.na(.)))
      
      # fit model with phenotype
      fit1 <- tryCatch({
        model <- lmerTest::lmer(as.formula(paste("value ~", "clean_reads_FQ_1 + dna_conc + BATCH_NUMBER +", lowest_p_phenotype, "+", phenotype, "* Timepoint_categorical + (1|NEXT_ID)")), data = df_filtered, REML = FALSE)
        list(success=TRUE, model=model, summary=broom::glance(model), tidied=broom::tidy(model) %>% mutate(corrected_phenotype = phenotype, factor = factor_name))
      }, error=function(e) {
        list(success=FALSE, error_message=conditionMessage(e), model=NULL, summary=NULL, tidied=NULL)
      }, warning = function(w) {
        list(success=FALSE, warning_message=conditionMessage(w), model=NULL, summary=NULL, tidied=NULL)
      })
      
      # fit model without phenotype
      fit0 <- tryCatch({
        model <- lmerTest::lmer(as.formula(paste("value ~", "clean_reads_FQ_1 + dna_conc + BATCH_NUMBER +", lowest_p_phenotype, "+ Timepoint_categorical + (1|NEXT_ID)")), data = df_filtered, REML = FALSE)
        list(success=TRUE, model=model, summary=broom::glance(model), tidied=broom::tidy(model) %>% mutate(corrected_phenotype = phenotype, factor = factor_name))
      }, error=function(e) {
        list(success=FALSE, error_message=conditionMessage(e), model=NULL, summary=NULL, tidied=NULL)
      }, warning = function(w) {
        list(success=FALSE, warning_message=conditionMessage(w), model=NULL, summary=NULL, tidied=NULL)
      })
      
      # compare models if both were successfully fitted
      if (fit1$success && fit0$success) {
        model_comparison <- anova(fit1$model, fit0$model, test = "LRT")
        model_comparison_tidied <- broom::tidy(model_comparison) %>% mutate(corrected_phenotype = phenotype, factor = factor_name)
      } else {
        model_comparison <- NULL
        model_comparison_tidied <- NULL
      }
      
      # store the result
      corrected_results[[phenotype]] <- list(fit1=fit1, fit0=fit0, model_comparison=model_comparison, model_comparison_tidied=model_comparison_tidied)
    }
  }
  
  # return results
  return(list(initial_results=initial_results, corrected_results=corrected_results, lowest_p_phenotype=lowest_p_phenotype))
}

# helper function to summarize model comparisons
summarize_model_comparisons <- function(results) {
  summary_df <- data.frame()
  
  for (i in seq_along(results)) {
    model_comparison_tidied <- results[[i]]$model_comparison_tidied
    
    if (!is.null(model_comparison_tidied)) {
      summary_df <- bind_rows(summary_df, model_comparison_tidied) %>% filter(!is.na(p.value)) %>%  mutate(p.adj = p.adjust(p.value, method = "bonferroni")) %>%
        select(npar,df,p.value,matches("pheno*"),factor,p.adj) %>% arrange(p.adj)
    }
  }
  
  return(summary_df)
}
```

## Run Models
This is fast and can be done on a personal machine.
```{r}
# Factor 1 ----
Fac1_models <-run_and_correct_MM(df = data_to_test, phenotypes = phenotypes_to_test, factor_name = "Factor1")

# summarize results
Fac1_initial_summary <- summarize_model_comparisons(Fac1_models$initial_results) %>% mutate(source = "initial_model")
Fac1_corrected_summary <- summarize_model_comparisons(Fac1_models$corrected_results) %>%   mutate(source = "corrected_model")
Fac1_combined_summary <- bind_rows(Fac1_initial_summary, Fac1_corrected_summary) %>%  mutate_all(~ ifelse(is.na(.), "", .)) %>% select(factor, source, npar, df, phenotype, corrected_phenotype, p.value, p.adj)

# table of results
kable(Fac1_combined_summary, format = "html", table.attr = "class='table table-bordered'")

# Factor 2 ----
Fac2_models <-run_and_correct_MM(df = data_to_test, phenotypes = phenotypes_to_test, factor_name = "Factor2")

# summarize results
Fac2_initial_summary <- summarize_model_comparisons(Fac2_models$initial_results) %>% mutate(source = "initial_model")
Fac2_corrected_summary <- summarize_model_comparisons(Fac2_models$corrected_results) %>%   mutate(source = "corrected_model")
Fac2_combined_summary <- bind_rows(Fac2_initial_summary, Fac2_corrected_summary) %>%  mutate_all(~ ifelse(is.na(.), "", .)) %>% select(factor, source, npar, df, phenotype, corrected_phenotype, p.value, p.adj)

# table of results
kable(Fac2_combined_summary, format = "html", table.attr = "class='table table-bordered'")


# Factor 3 ----
Fac3_models <-run_and_correct_MM(df = data_to_test, phenotypes = phenotypes_to_test, factor_name = "Factor3")

# summarize results
Fac3_initial_summary <- summarize_model_comparisons(Fac3_models$initial_results) %>% mutate(source = "initial_model")
Fac3_corrected_summary <- summarize_model_comparisons(Fac3_models$corrected_results) %>%   mutate(source = "corrected_model")
Fac3_combined_summary <- bind_rows(Fac3_initial_summary, Fac3_corrected_summary) %>%  mutate_all(~ ifelse(is.na(.), "", .)) %>% select(factor, source, npar, df, phenotype, corrected_phenotype, p.value, p.adj)

# table of results
kable(Fac3_combined_summary, format = "html", table.attr = "class='table table-bordered'")

# Factor 4 ----
Fac4_models <-run_and_correct_MM(df = data_to_test, phenotypes = phenotypes_to_test, factor_name = "Factor4")

# summarize results
Fac4_initial_summary <- summarize_model_comparisons(Fac4_models$initial_results) %>% mutate(source = "initial_model")
Fac4_corrected_summary <- summarize_model_comparisons(Fac4_models$corrected_results) %>%   mutate(source = "corrected_model")
Fac4_combined_summary <- bind_rows(Fac4_initial_summary, Fac4_corrected_summary) %>%  mutate_all(~ ifelse(is.na(.), "", .)) %>% select(factor, source, npar, df, phenotype, corrected_phenotype, p.value, p.adj)

# table of results
kable(Fac4_combined_summary, format = "html", table.attr = "class='table table-bordered'")

# Factor 5 ----
Fac5_models <-run_and_correct_MM(df = data_to_test, phenotypes = phenotypes_to_test, factor_name = "Factor5")

# summarize results
Fac5_initial_summary <- summarize_model_comparisons(Fac5_models$initial_results) %>% mutate(source = "initial_model")
Fac5_corrected_summary <- summarize_model_comparisons(Fac5_models$corrected_results) %>%   mutate(source = "corrected_model")
Fac5_combined_summary <- bind_rows(Fac5_initial_summary, Fac5_corrected_summary) %>%  mutate_all(~ ifelse(is.na(.), "", .)) %>% select(factor, source, npar, df, phenotype, corrected_phenotype, p.value, p.adj)

# table of results
kable(Fac5_combined_summary, format = "html", table.attr = "class='table table-bordered'")

# Factor 6 ----
Fac6_models <-run_and_correct_MM(df = data_to_test, phenotypes = phenotypes_to_test, factor_name = "Factor6")

# summarize results
Fac6_initial_summary <- summarize_model_comparisons(Fac6_models$initial_results) %>% mutate(source = "initial_model")
Fac6_corrected_summary <- summarize_model_comparisons(Fac6_models$corrected_results) %>%   mutate(source = "corrected_model")
Fac6_combined_summary <- bind_rows(Fac6_initial_summary, Fac6_corrected_summary) %>%  mutate_all(~ ifelse(is.na(.), "", .)) %>% select(factor, source, npar, df, phenotype, corrected_phenotype, p.value, p.adj)

# table of results
kable(Fac6_combined_summary, format = "html", table.attr = "class='table table-bordered'")

# Factor 7 ----
Fac7_models <-run_and_correct_MM(df = data_to_test, phenotypes = phenotypes_to_test, factor_name = "Factor7")

# summarize results
Fac7_initial_summary <- summarize_model_comparisons(Fac7_models$initial_results) %>% mutate(source = "initial_model")
Fac7_corrected_summary <- summarize_model_comparisons(Fac7_models$corrected_results) %>%   mutate(source = "corrected_model")
Fac7_combined_summary <- bind_rows(Fac7_initial_summary, Fac7_corrected_summary) %>%  mutate_all(~ ifelse(is.na(.), "", .)) %>% select(factor, source, npar, df, phenotype, corrected_phenotype, p.value, p.adj)

# table of results
kable(Fac7_combined_summary, format = "html", table.attr = "class='table table-bordered'")

# Factor 8 ----
Fac8_models <-run_and_correct_MM(df = data_to_test, phenotypes = phenotypes_to_test, factor_name = "Factor8")

# summarize results
Fac8_initial_summary <- summarize_model_comparisons(Fac8_models$initial_results) %>% mutate(source = "initial_model")
Fac8_corrected_summary <- summarize_model_comparisons(Fac8_models$corrected_results) %>%   mutate(source = "corrected_model")
Fac8_combined_summary <- bind_rows(Fac8_initial_summary, Fac8_corrected_summary) %>%  mutate_all(~ ifelse(is.na(.), "", .)) %>% select(factor, source, npar, df, phenotype, corrected_phenotype, p.value, p.adj)

# table of results
kable(Fac8_combined_summary, format = "html", table.attr = "class='table table-bordered'")

# Factor 9 ----
Fac9_models <-run_and_correct_MM(df = data_to_test, phenotypes = phenotypes_to_test, factor_name = "Factor9")

# summarize results
Fac9_initial_summary <- summarize_model_comparisons(Fac9_models$initial_results) %>% mutate(source = "initial_model")
Fac9_corrected_summary <- summarize_model_comparisons(Fac9_models$corrected_results) %>%   mutate(source = "corrected_model")
Fac9_combined_summary <- bind_rows(Fac9_initial_summary, Fac9_corrected_summary) %>%  mutate_all(~ ifelse(is.na(.), "", .)) %>% select(factor, source, npar, df, phenotype, corrected_phenotype, p.value, p.adj)

# table of results
kable(Fac9_combined_summary, format = "html", table.attr = "class='table table-bordered'")

# Factor 10 ----
Fac10_models <-run_and_correct_MM(df = data_to_test, phenotypes = phenotypes_to_test, factor_name = "Factor10")

# summarize results
Fac10_initial_summary <- summarize_model_comparisons(Fac10_models$initial_results) %>% mutate(source = "initial_model")
Fac10_corrected_summary <- summarize_model_comparisons(Fac10_models$corrected_results) %>%   mutate(source = "corrected_model")
Fac10_combined_summary <- bind_rows(Fac10_initial_summary, Fac10_corrected_summary) %>%  mutate_all(~ ifelse(is.na(.), "", .)) %>% select(factor, source, npar, df, phenotype, corrected_phenotype, p.value, p.adj)

# table of results
kable(Fac10_combined_summary, format = "html", table.attr = "class='table table-bordered'")
```


## Table of Results
```{r}
# put it all together
Fac_summaries_vec <- paste0("Fac", 1:10, "_combined_summary")

# Access each variable dynamically and store in a list
summaries_list <- lapply(Fac_summaries_vec, get)

# final table of results
final_table <-
bind_rows(summaries_list) %>% 
    mutate(phenotype =
           case_when(phenotype == "birth_birthcard_total_duration_delivery_min" ~ "delivery_duration",
                     phenotype == "birth_deliverybirthcard_mode_binary" ~ "birth_mode",
                     phenotype == "infant_birthcard_feeding_mode_after_birth" ~ "feeding_mode",
                     phenotype == "birth_deliverybirthcard_place_delivery_simple" ~ "delivery_place",
                     phenotype == "mother_birthcard_parity" ~ "parity",
                     phenotype == "family_pets_any" ~ "pets",
                     phenotype == "infant_health_eczema_diagnosis_strict" ~ "eczema",
                     phenotype == "infant_growth_birth_weight_kg" ~ "birth_weight",
                     phenotype == "infant_health_food_allergy" ~ "food_allergy",
                     phenotype == "infant_misc_sex" ~ "sex",
                     phenotype == "mother_birthcard_age_at_delivery" ~ "mother_age",
                     phenotype == "mother_birthcard_duration_water_broken_min" ~ "water_broken_duration",
                     phenotype == "mother_health_smoked_one_whole_year_p18" ~ "smoked",
                     TRUE ~ phenotype)) %>%
    mutate(corrected_phenotype =
           case_when(corrected_phenotype == "birth_birthcard_total_duration_delivery_min" ~ "delivery_duration",
                     corrected_phenotype == "birth_deliverybirthcard_mode_binary" ~ "birth_mode",
                     corrected_phenotype == "infant_birthcard_feeding_mode_after_birth" ~ "feeding_mode",
                     corrected_phenotype == "birth_deliverybirthcard_place_delivery_simple" ~ "delivery_place",
                     corrected_phenotype == "mother_birthcard_parity" ~ "parity",
                     corrected_phenotype == "family_pets_any" ~ "pets",
                     corrected_phenotype == "infant_health_eczema_diagnosis_strict" ~ "eczema",
                     corrected_phenotype == "infant_growth_birth_weight_kg" ~ "birth_weight",
                     corrected_phenotype == "infant_health_food_allergy" ~ "food_allergy",
                     corrected_phenotype == "infant_misc_sex" ~ "sex",
                     corrected_phenotype == "mother_birthcard_age_at_delivery" ~ "mother_age",
                     corrected_phenotype == "mother_birthcard_duration_water_broken_min" ~ "water_broken_duration",
                     corrected_phenotype == "mother_health_smoked_one_whole_year_p18" ~ "smoked",
                     TRUE ~ corrected_phenotype)) %>%
  group_by(factor, source) %>%
  filter(p.adj < 0.05)

# create latex table
kable(final_table, "latex", booktabs = TRUE) %>%
  kable_styling(latex_options = c("hold_position")) %>%
  pack_rows("Factor1", 1, 10, bold = TRUE, italic = TRUE) %>%
  pack_rows("Factor2", 11, 16, bold = TRUE, italic = TRUE) %>%
  pack_rows("Factor3", 17, 21, bold = TRUE, italic = TRUE) %>%
  pack_rows("Factor4", 22, 23, bold = TRUE, italic = TRUE) %>%
  pack_rows("Factor5", 24, 24, bold = TRUE, italic = TRUE) %>%
  pack_rows("Factor6", 25, 25, bold = TRUE, italic = TRUE) %>%
  pack_rows("Factor8", 26, 26, bold = TRUE, italic = TRUE) %>%
  pack_rows("Factor10", 27, 29, bold = TRUE, italic = TRUE)
```

## Examine Top Associations Per Factor
```{r}
# helper function to search through models
find_matching_model_initial <- function(model_results, search_phenotype, search_factor) {
  for (i in seq_along(model_results[[1]])) {
    # access the model comparison tidied tibble
    model_comparison_tidied <- model_results[[1]][[i]]$model_comparison_tidied
    
    if (!is.null(model_comparison_tidied)) {
      # check if there is a match for the given phenotype and factor
      match <- model_comparison_tidied %>%
        filter(phenotype == search_phenotype, factor == search_factor)
      
      if (nrow(match) > 0) {
        return(
          list(tib = 
                 model_results[[1]][[i]]$fit1$tidied %>%
                 select(factor, term, df, estimate, std.error, p.value) %>%
                 mutate(term = str_replace_all(term, "Timepoint_categorical", "")) %>%
                 filter(term != "sd__(Intercept)") %>%
                 filter(term != "sd__Observation"),
                 # mutate(term = str_replace_all(term, "birth_birthcard_total_duration_delivery_min", "delivery_duration") %>%
                 #          str_replace_all("birth_deliverybirthcard_mode_binary", "birth_mode") %>%
                 #          str_replace_all("infant_birthcard_feeding_mode_after_birth", "feeding_mode") %>%
                 #          str_replace_all("birth_deliverybirthcard_place_delivery_simple", "delivery_place") %>%
                 #          str_replace_all("mother_birthcard_parity", "parity") %>%
                 #          str_replace_all("family_pets_any", "pets") %>%
                 #          str_replace_all("infant_health_eczema_diagnosis_strict", "eczema") %>%
                 #          str_replace_all("infant_growth_birth_weight_kg", "birth_weight") %>%
                 #          str_replace_all("infant_health_food_allergy", "food_allergy") %>%
                 #          str_replace_all("infant_misc_sex", "sex") %>%
                 #          str_replace_all("mother_birthcard_age_at_delivery", "mother_age") %>%
                 #          str_replace_all("mother_birthcard_duration_water_broken_min", "water_broken_duration") %>%
                 #          str_replace_all("mother_health_smoked_one_whole_year_p18", "smoked")
               
               
               table =  model_results[[1]][[i]]$fit1$tidied %>%
                 select(factor, term, df, estimate, p.value) %>%
                 mutate(term = str_replace_all(term, "Timepoint_categorical", "")) %>%
                 # mutate(term = str_replace_all(term, "birth_birthcard_total_duration_delivery_min", "delivery_duration") %>%
                 #          str_replace_all("birth_deliverybirthcard_mode_binary", "birth_mode") %>%
                 #          str_replace_all("infant_birthcard_feeding_mode_after_birth", "feeding_mode") %>%
                 #          str_replace_all("birth_deliverybirthcard_place_delivery_simple", "delivery_place") %>%
                 #          str_replace_all("mother_birthcard_parity", "parity") %>%
                 #          str_replace_all("family_pets_any", "pets") %>%
                 #          str_replace_all("infant_health_eczema_diagnosis_strict", "eczema") %>%
                 #          str_replace_all("infant_growth_birth_weight_kg", "birth_weight") %>%
                 #          str_replace_all("infant_health_food_allergy", "food_allergy") %>%
                 #          str_replace_all("infant_misc_sex", "sex") %>%
                 #          str_replace_all("mother_birthcard_age_at_delivery", "mother_age") %>%
                 #          str_replace_all("mother_birthcard_duration_water_broken_min", "water_broken_duration") %>%
                 #          str_replace_all("mother_health_smoked_one_whole_year_p18", "smoked")) %>% 
                 filter(!term %in% c("sd__(Intercept)", "sd__Observation")) 
                 # kable(format = "latex", caption = "Model Results", booktabs = TRUE) %>%
                 # kable_styling(latex_options = c( "hold_position"))
          )
        )
        #return(model_results[[1]][[i]])
      }
    }
  }
  
  return(NULL)  # Return NULL if no match is found
}

find_matching_model_corrected <- function(model_results, search_phenotype, search_factor) {
  for (i in seq_along(model_results[[2]])) {
    # access the model comparison tidied tibble
    model_comparison_tidied <- model_results[[2]][[i]]$model_comparison_tidied
    
    if (!is.null(model_comparison_tidied)) {
      # check if there is a match for the given phenotype and factor
      match <- model_comparison_tidied %>%
        filter(corrected_phenotype == search_phenotype, factor == search_factor)
      
      if (nrow(match) > 0) { return(
          list(tib = 
          model_results[[2]][[i]]$fit1$tidied %>%
          select(factor, term, df, estimate, std.error, p.value) %>%
            mutate(term = str_replace_all(term, "Timepoint_categorical", "")) %>%
            filter(term != "sd__(Intercept)") %>%
            filter(term != "sd__Observation"),
            # mutate(term = str_replace_all(term, "birth_birthcard_total_duration_delivery_min", "delivery_duration") %>%
            #  str_replace_all("birth_deliverybirthcard_mode_binary", "birth_mode") %>%
            #  str_replace_all("infant_birthcard_feeding_mode_after_birth", "feeding_mode") %>%
            #  str_replace_all("birth_deliverybirthcard_place_delivery_simple", "delivery_place") %>%
            #  str_replace_all("mother_birthcard_parity", "parity") %>%
            #  str_replace_all("family_pets_any", "pets") %>%
            #  str_replace_all("infant_health_eczema_diagnosis_strict", "eczema") %>%
            #  str_replace_all("infant_growth_birth_weight_kg", "birth_weight") %>%
            #  str_replace_all("infant_health_food_allergy", "food_allergy") %>%
            #  str_replace_all("infant_misc_sex", "sex") %>%
            #  str_replace_all("mother_birthcard_age_at_delivery", "mother_age") %>%
            #  str_replace_all("mother_birthcard_duration_water_broken_min", "water_broken_duration") %>%
            #  str_replace_all("mother_health_smoked_one_whole_year_p18", "smoked")),
          
          table =  model_results[[2]][[i]]$fit1$tidied %>%
          select(factor, term, df, estimate, p.value) %>%
            mutate(term = str_replace_all(term, "Timepoint_categorical", "")) %>%
            mutate(term = str_replace_all(term, "birth_birthcard_total_duration_delivery_min", "delivery_duration") %>%
             str_replace_all("birth_deliverybirthcard_mode_binary", "birth_mode") %>%
             str_replace_all("infant_birthcard_feeding_mode_after_birth", "feeding_mode") %>%
             str_replace_all("birth_deliverybirthcard_place_delivery_simple", "delivery_place") %>%
             str_replace_all("mother_birthcard_parity", "parity") %>%
             str_replace_all("family_pets_any", "pets") %>%
             str_replace_all("infant_health_eczema_diagnosis_strict", "eczema") %>%
             str_replace_all("infant_growth_birth_weight_kg", "birth_weight") %>%
             str_replace_all("infant_health_food_allergy", "food_allergy") %>%
             str_replace_all("infant_misc_sex", "sex") %>%
             str_replace_all("mother_birthcard_age_at_delivery", "mother_age") %>%
             str_replace_all("mother_birthcard_duration_water_broken_min", "water_broken_duration") %>%
             str_replace_all("mother_health_smoked_one_whole_year_p18", "smoked")) %>% 
            filter(!term %in% c("sd__(Intercept)", "sd__Observation")) %>%
            kable(format = "latex", caption = "Model Results", booktabs = TRUE) %>%
            kable_styling(latex_options = c( "hold_position"))
          )
  )
      }
    }
  }
  
  return(NULL)  # Return NULL if no match is found
}

## Factor 1
mod_Fac1 <- find_matching_model_initial(Fac1_models, "birth_deliverybirthcard_mode_binary", "Factor1")

## Factor 2
mod_Fac2 <- find_matching_model_initial(Fac2_models, "birth_deliverybirthcard_mode_binary", "Factor2")

## Factor 3
mod_Fac3 <- find_matching_model_initial(Fac3_models, "infant_birthcard_feeding_mode_after_birth", "Factor3")

## Factor 4
mod_Fac4 <- find_matching_model_initial(Fac4_models, "mother_health_smoked_one_whole_year_p18", "Factor4")

## Factor 5
mod_Fac5 <- find_matching_model_initial(Fac5_models, "birth_deliverybirthcard_place_delivery_simple", "Factor5")

## Factor 6
mod_Fac6 <- find_matching_model_initial(Fac6_models, "infant_birthcard_feeding_mode_after_birth", "Factor6")

## Factor 8
mod_Fac8 <- find_matching_model_initial(Fac8_models, "infant_birthcard_feeding_mode_after_birth", "Factor8")

## Factor 10
mod_Fac10 <- find_matching_model_corrected(Fac10_models, "mother_birthcard_parity", "Factor10") # this is where the corrected phenotype had a lower p-value the the most significant association in the first series of models.


model_table <- bind_rows(mod_Fac1$tib, mod_Fac2$tib, mod_Fac3$tib, mod_Fac4$tib, mod_Fac5$tib, mod_Fac6$tib, mod_Fac8$tib, mod_Fac10$tib) %>% print(n = nrow(.))

write.xlsx(model_table, file = "model_results_per_factor.xlsx", asTable = TRUE)
```

# Examine the Relationship Between Factors and Diversity.


## Clear Env and Load Libraries and Data
```{r, warning=FALSE, message = FALSE}
# clear environment
rm(list = ls())

# load libraries
library(kableExtra)
library(lme4)
library(lmerTest)
library(broom)
library(broom.mixed)
library(tidyverse)
library(MOFA2)
library(here)
library(MuMIn)
library(partR2)
library(patchwork)
```

## Prep Data
```{r}
# load MEFISTO results ----
MOFAobject <- load_model(here("LLNEXT_microbiome_model_26_01_2024.hdf5"), load_interpol_Z = FALSE)

# prep diversity data----
## calc diversity (from Trishla's script: LIFELINES_NEXT_gut_microbiome_filtering_CLR_transformation_metaphlan_4_infants.R)
metadata<-read.csv(file=here("data_meta/LLNEXT_metadata_CLEAN_10_07_2023.txt"), header = TRUE, sep = "\t")
taxa<-read.csv(file=here("data_sequencing/LLNEXT_metaphlan_4_CLEAN_10_07_2023.txt"), header = TRUE, sep = "\t")

##1) Infants 
infant_metadata<-metadata[metadata$Type=="infant",]
infant_taxa<-taxa[row.names(taxa)%in% rownames(infant_metadata),] 
infant_taxa<-infant_taxa[match(row.names(infant_taxa),row.names(infant_metadata)),]
infant_NEXT_ID<-infant_metadata %>% select(NEXT_ID)
infant_taxa_all<-merge(infant_NEXT_ID,infant_taxa, by="row.names" )
row.names(infant_taxa_all)<-infant_taxa_all$Row.names
infant_taxa_all$Row.names=NULL
unique_counts <- sapply(infant_taxa_all, function(x) length(unique(infant_taxa_all$NEXT_ID[x >0.1]))) # Here I am counting the unique elements in the NEXT_ID column where the corresponding value in each column (i.e., x) is greater than the given cut-off. 
infant_taxa_all_filt <- infant_taxa_all[, unique_counts >= 0.1*length(unique(infant_taxa_all$NEXT_ID)) ] # Setting a 10% cut-off on prevalence 
infant_taxa_all_filt$NEXT_ID=NULL
infant_taxa_all_filt$UNCLASSIFIED=NULL
infant_taxa_all_filt[1:5,1:5]

infant_taxa_SGB_filt=infant_taxa_all_filt[,grep("t__",colnames(infant_taxa_all_filt))]

## taxa with metadata
data_meta_with_NG_ID <-
  infant_taxa_SGB_filt %>%
  mutate(NG_ID = row.names(infant_taxa_SGB_filt)) %>%
  left_join(., infant_metadata, by = "NG_ID") %>% 
  select(106:132) %>%
  rename(sample = NG_ID)

## get factor scores
z_scores <- as_tibble(MOFAobject@expectations$Z$single_group)
z_scores$sample <- rownames(MOFAobject@expectations$Z$single_group)

# calc div
data_diversity <-
  infant_taxa_SGB_filt %>%
  mutate(shannon = vegan::diversity(across(), index = "shannon"),
         richness = specnumber(across())) %>%
  mutate(NG_ID = row.names(.)) %>%
  select(NG_ID, shannon, richness) %>%
  rename(sample = NG_ID)

# join with factors and metadata
data_diversity_factors <-
  data_diversity %>%
  right_join(., z_scores, by = "sample") 

#all data
data_diversity_factors %>% 
  left_join(.,data_meta_with_NG_ID, by = "sample")

# load data used for MEFISTO ----
load(here("data_for_MOFA_26_01_2024.rda"))
head(data_for_MOFA)

# inverse rank transformation func
inverse_rank_transform <- function(x) {
  ranks <- rank(x, na.last = "keep", ties.method = "average") # rank data
  qnorm((ranks - 0.5) / length(ranks))          # apply transformation
}

data_meta <- read.csv(file=here("data_meta/LLNEXT_metadata_03_08_2023.txt"), header = TRUE, sep = "\t")

data_meta2<-
data_meta %>%
  filter(Type == "infant") %>%
  select(NG_ID, NEXT_ID, dna_conc, clean_reads_FQ_1, BATCH_NUMBER,Timepoint_categorical) %>% 
  mutate(across(-c(NG_ID, Timepoint_categorical, NEXT_ID), inverse_rank_transform)) %>%
  rename(sample = NG_ID)



# make vector of phenotypes (I'm calling them phenotypes here, but i really mean covariates. why? i'm reusing old code)
phenotypes_and_covariates <- names(data_meta2)[c(3:5)]
phenotypes_to_test <- phenotypes_and_covariates

data_factors <-
  get_factors(MOFAobject, as.data.frame = TRUE, factors = c(1:10)) 

data_to_test <-
  data_meta2 %>% 
  left_join(., data_factors, by = "sample") %>%
  ungroup() %>%
  left_join(., data_diversity_factors, by = "sample")
```

## Formulate and Run Models
```{r}
# Model Function
#Combined function to run mixed models. An initial set of associations is run. Then a second set of associations is run, correcting for the most significant phenotype in the initial set of associations. 

# function to run mixed models with stepwise correction 
run_and_correct_MM <- function(df, factor_name) {
    
    # remove NAs from data and select needed data 
    df_filtered <- df %>%
      ungroup() %>%
      filter(factor == factor_name) %>%
      select(shannon, value, clean_reads_FQ_1, dna_conc, BATCH_NUMBER, Timepoint_categorical, NEXT_ID) %>%
      mutate(Timepoint_categorical = factor(Timepoint_categorical, levels = c("W2", "M1", "M2","M3","M6","M9","M12"))) %>%
      filter(if_all(everything(), ~ !is.na(.)))
    
    # fit model with phenotype
    fit1 <- tryCatch({
      model <- lmerTest::lmer(as.formula(paste("shannon ~", "value + clean_reads_FQ_1 + dna_conc + BATCH_NUMBER + value * Timepoint_categorical + (1|NEXT_ID)")), data = df_filtered, REML = FALSE)
      list(success=TRUE, model=model, summary=broom::glance(model), tidied=broom::tidy(model) %>% mutate(factor = factor_name))
    }, error=function(e) {
      list(success=FALSE, error_message=conditionMessage(e), model=NULL, summary=NULL, tidied=NULL)
    }, warning = function(w) {
      list(success=FALSE, warning_message=conditionMessage(w), model=NULL, summary=NULL, tidied=NULL)
    })
    
    # fit model without phenotype
    fit0 <- tryCatch({
      model <- lmerTest::lmer(as.formula(paste("shannon ~", "clean_reads_FQ_1 + dna_conc + BATCH_NUMBER + Timepoint_categorical + (1|NEXT_ID)")), data = df_filtered, REML = FALSE)
      list(success=TRUE, model=model, summary=broom::glance(model), tidied=broom::tidy(model) %>% mutate(factor = factor_name))
    }, error=function(e) {
      list(success=FALSE, error_message=conditionMessage(e), model=NULL, summary=NULL, tidied=NULL)
    }, warning = function(w) {
      list(success=FALSE, warning_message=conditionMessage(w), model=NULL, summary=NULL, tidied=NULL)
    })
    
    # compare models if both were successfully fitted
    if (fit1$success && fit0$success) {
      model_comparison <- anova(fit1$model, fit0$model, test = "LRT")
      model_comparison_tidied <- broom::tidy(model_comparison) %>% mutate(factor = factor_name)
    } else {
      model_comparison <- NULL
      model_comparison_tidied <- NULL
    }
    
    # store all results 
    res <- list(fit1=fit1, fit0=fit0, model_comparison=model_comparison, model_comparison_tidied=model_comparison_tidied)
  
    return(res)
}


# Run Models
# Factor 1 ----
Fac1_models <-run_and_correct_MM(df = data_to_test, factor_name = "Factor1")
Fac2_models <-run_and_correct_MM(df = data_to_test, factor_name = "Factor2")
Fac3_models <-run_and_correct_MM(df = data_to_test, factor_name = "Factor3")
Fac4_models <-run_and_correct_MM(df = data_to_test, factor_name = "Factor4")
Fac5_models <-run_and_correct_MM(df = data_to_test, factor_name = "Factor5")
Fac6_models <-run_and_correct_MM(df = data_to_test, factor_name = "Factor6")
Fac7_models <-run_and_correct_MM(df = data_to_test, factor_name = "Factor7")
Fac8_models <-run_and_correct_MM(df = data_to_test, factor_name = "Factor8")
Fac9_models <-run_and_correct_MM(df = data_to_test, factor_name = "Factor9")
Fac10_models <-run_and_correct_MM(df = data_to_test, factor_name = "Factor10")
```

## Examine First Three Factors
```{r}
# load data phenotypes to add to plots at meta info
source(here("prep_phenotypes.R")) # script that prep phenotypes
data_phenos <-
  phenos %>%
  rename(sample = NG_ID) %>%
  select(sample, NEXT_ID, birth_deliverybirthcard_mode_binary,infant_birthcard_feeding_mode_after_birth)

data_to_test <-
data_to_test %>%
  left_join(., data_phenos, by = "sample")

# Examine Factor 1----

## extract model info
model_summary <- summary(Fac1_models$fit1$model)
intercept <- model_summary$coefficients[1,1]
slope <- model_summary$coefficients[2,1]
R2 <- r.squaredGLMM(Fac1_models$fit1$model)
pval <- Fac1_models$model_comparison_tidied$p.value[2]

# variance of null and test model
full_residual_var <- attr(VarCorr(Fac1_models$fit1$model), "sc")^2
reduced_residual_var <- attr(VarCorr(Fac1_models$fit0$model), "sc")^2

# variance explained by the factor
var_explained <- 1 - (full_residual_var / reduced_residual_var)

#colors
col_vec <- c("#FFC20A", "#0C7BDC")
col_vec2 <- c("#FFC20A", "#0C7BDC","#004D40")

# plot
p_Fac_shannon <-
  ggplot(data_to_test %>% filter(!is.na(birth_deliverybirthcard_mode_binary))) +
  geom_point(aes_string(x = "Factor1", y = "shannon", group = "birth_deliverybirthcard_mode_binary", color = "birth_deliverybirthcard_mode_binary")) +
  scale_color_manual(values=col_vec) +
 annotate("text", x = -Inf, y = Inf, 
           label = bquote(R^2 == .(round(var_explained, 3)) ~ ", " ~ p == .(format.pval(pval, digits = 3))),
           hjust = -0.1, vjust = 1, size = 4, color = "black") +
  geom_abline(slope = slope, intercept = intercept, color = "black") +
  theme_minimal()

p_Fac_shannon

bp_shannon_fac <-
  ggplot(data_to_test %>% filter(!is.na(birth_deliverybirthcard_mode_binary)) %>% mutate(birth_deliverybirthcard_mode_binary = fct_relevel(birth_deliverybirthcard_mode_binary, "CS", "VG"))) +
  geom_boxplot(aes_string(x = "birth_deliverybirthcard_mode_binary", y = "Factor1", fill = "birth_deliverybirthcard_mode_binary")) +
  #geom_point() +
  #geom_abline(slope = slope, intercept = intercept, color = "blue") +
  coord_flip() +
  scale_fill_manual(values = col_vec) +
  theme_minimal() +
  labs(x = NULL)

bp_shannon_fac


bp_bm_shannon <-
  ggplot(data_to_test %>% filter(!is.na(birth_deliverybirthcard_mode_binary)) %>% mutate(birth_deliverybirthcard_mode_binary = fct_relevel(birth_deliverybirthcard_mode_binary, "CS", "VG"))) +
  geom_boxplot(aes_string(x = "birth_deliverybirthcard_mode_binary", y = "shannon", fill = "birth_deliverybirthcard_mode_binary")) +
  scale_fill_manual(values = col_vec) +
  #geom_point() +
  #geom_abline(slope = slope, intercept = intercept, color = "blue") +
  theme_minimal() 
      

fac1_div_fac <-
bp_shannon_fac + 
  plot_spacer() + 
  p_Fac_shannon + 
  bp_bm_shannon +
  plot_layout(ncol = 2, nrow = 2, widths = c(4, 1), heights = c(1, 4))

# Examine Factor 2----

## extract model info
model_summary <- summary(Fac2_models$fit1$model)
intercept <- model_summary$coefficients[1,1]
slope <- model_summary$coefficients[2,1]
R2 <- r.squaredGLMM(Fac2_models$fit1$model)
pval <- Fac1_models$model_comparison_tidied$p.value[2]

# variance of null and test model
full_residual_var <- attr(VarCorr(Fac2_models$fit1$model), "sc")^2
reduced_residual_var <- attr(VarCorr(Fac2_models$fit0$model), "sc")^2

# variance explained by the factor
var_explained <- 1 - (full_residual_var / reduced_residual_var)

# plot

p_Fac_shannon <-
ggplot(data_to_test%>% filter(!is.na(birth_deliverybirthcard_mode_binary))) +
  geom_point(aes_string(x = "Factor2", y = "shannon", group = "birth_deliverybirthcard_mode_binary", color = "birth_deliverybirthcard_mode_binary")) +
   scale_color_manual(values=col_vec) +
 annotate("text", x = -Inf, y = Inf, 
           label = bquote(R^2 == .(round(var_explained, 3)) ~ ", " ~ p == .(format.pval(pval, digits = 3))),
           hjust = -0.1, vjust = 1, size = 4, color = "black") +
  geom_abline(slope = slope, intercept = intercept, color = "black") +
  geom_abline(slope = slope, intercept = intercept, color = "black") +
  theme_minimal()

bp_shannon_fac <-
ggplot(data_to_test %>% filter(!is.na(birth_deliverybirthcard_mode_binary)) %>% mutate(birth_deliverybirthcard_mode_binary = fct_relevel(birth_deliverybirthcard_mode_binary, "CS", "VG"))) +
  geom_boxplot(aes_string(x = "birth_deliverybirthcard_mode_binary", y = "Factor2", fill = "birth_deliverybirthcard_mode_binary")) +
  #geom_point() +
  #geom_abline(slope = slope, intercept = intercept, color = "blue") +
  scale_fill_manual(values = col_vec) +
  coord_flip() +
  theme_minimal() +
  labs(x = NULL)

bp_bm_shannon <-
ggplot(data_to_test %>% filter(!is.na(birth_deliverybirthcard_mode_binary)) %>% mutate(birth_deliverybirthcard_mode_binary = fct_relevel(birth_deliverybirthcard_mode_binary, "CS", "VG"))) +
  geom_boxplot(aes_string(x = "birth_deliverybirthcard_mode_binary", y = "shannon", fill = "birth_deliverybirthcard_mode_binary")) +
  scale_fill_manual(values = col_vec) +
  #geom_point() +
  #geom_abline(slope = slope, intercept = intercept, color = "blue") +
  theme_minimal()

fac2_div_fac <-
bp_shannon_fac + 
  plot_spacer() + 
  p_Fac_shannon + 
  bp_bm_shannon +
  plot_layout(ncol = 2, nrow = 2, widths = c(4, 1), heights = c(1, 4))


# Examine Factor 3----

## extract model info
model_summary <- summary(Fac3_models$fit1$model)
intercept <- model_summary$coefficients[1,1]
slope <- model_summary$coefficients[2,1]
R2 <- r.squaredGLMM(Fac3_models$fit1$model)
pval <- Fac1_models$model_comparison_tidied$p.value[2]

# variance of null and test model
full_residual_var <- attr(VarCorr(Fac3_models$fit1$model), "sc")^2
reduced_residual_var <- attr(VarCorr(Fac3_models$fit0$model), "sc")^2

# variance explained by the factor
var_explained <- 1 - (full_residual_var / reduced_residual_var)

# plot

p_Fac_shannon <-
  ggplot(data_to_test %>% filter(!is.na(infant_birthcard_feeding_mode_after_birth)) %>% mutate(infant_birthcard_feeding_mode_after_birth = fct_relevel(infant_birthcard_feeding_mode_after_birth, "BF", "MF", "FF"))) +
  geom_point(aes_string(x = "Factor3", y = "shannon", color = "infant_birthcard_feeding_mode_after_birth")) +
   annotate("text", x = -Inf, y = Inf, 
           label = bquote(R^2 == .(round(var_explained, 3)) ~ ", " ~ p == .(format.pval(pval, digits = 3))),
           hjust = -0.1, vjust = 1, size = 4, color = "black") +
  geom_abline(slope = slope, intercept = intercept, color = "black") +
  scale_color_manual(values = col_vec2) +
  theme_minimal()

bp_shannon_fac <-
  data_to_test %>% 
  filter(!is.na(infant_birthcard_feeding_mode_after_birth)) %>% 
  mutate(infant_birthcard_feeding_mode_after_birth = fct_relevel(infant_birthcard_feeding_mode_after_birth, "BF", "MF", "FF")) %>% 
  ggplot() +
  geom_boxplot(aes_string(x = "infant_birthcard_feeding_mode_after_birth", y = "Factor3", fill = "infant_birthcard_feeding_mode_after_birth")) +
  scale_fill_manual(values = col_vec2) +
  coord_flip() +
  theme_minimal() +
  labs(x = NULL)

bp_bm_shannon <-
  data_to_test %>% 
  filter(!is.na(infant_birthcard_feeding_mode_after_birth)) %>% 
  mutate(infant_birthcard_feeding_mode_after_birth = fct_relevel(infant_birthcard_feeding_mode_after_birth, "BF", "MF", "FF")) %>% 
  ggplot() +
  geom_boxplot(aes_string(x = "infant_birthcard_feeding_mode_after_birth", y = "shannon", fill = "infant_birthcard_feeding_mode_after_birth")) +
  scale_fill_manual(values = col_vec2) +
  #geom_point() +
  #geom_abline(slope = slope, intercept = intercept, color = "blue") +
  theme_minimal()


fac3_div_fac <-
bp_shannon_fac + 
  plot_spacer() + 
  p_Fac_shannon + 
  bp_bm_shannon +
  plot_layout(ncol = 2, nrow = 2, widths = c(4, 1), heights = c(1, 4))

#view all plots

fac1_div_fac
fac2_div_fac
fac3_div_fac

# save plot data
saveRDS(fac1_div_fac, file = "fac1_div_fac.rds")
saveRDS(fac2_div_fac, file = "fac2_div_fac.rds")
saveRDS(fac3_div_fac, file = "fac3_div_fac.rds")

# save plot figure
ggsave("fac1_div_fac.png", plot = fac1_div_fac, width = 11.5, height = 5)
ggsave("fac2_div_fac.png", plot = fac2_div_fac, width = 11.5, height = 5)
ggsave("fac3_div_fac.png", plot = fac3_div_fac, width = 11.5, height = 5)
```



